src/SBInterpolatedImage.cpp

```diff
@@ -900,6 +782,198 @@ namespace galsim {
+
+        _ktab = boost::shared_ptr<KTable>(new KTable(Nk, _dk));
+        _maxk = Ninitial/2 * _dk;
+        dbg<<"_dk = "<<_dk<<std::endl;
+        dbg<<"_maxk = "<<_maxk<<std::endl;
+
+        // Only need to fill in x>=0 since the negative x's are the Hermitian
+        // conjugates of the positive x's.
+        int kxStart = 0;
+        int ikxStart = (realKImage.getXMin()+realKImage.getXMax()+1)/2;
+        int ky = -((realKImage.getYMax()-realKImage.getYMin()+1)/2);
+        dbg<<"kxStart = "<<kxStart<<", kyStart = "<<ky<<std::endl;
+        for (int iky = realKImage.getYMin(); iky<= realKImage.getYMax(); ++iky, ++ky) {
+             int kx = kxStart;
+             for (int ikx = ikxStart; ikx<= realKImage.getXMax(); ++ikx, ++kx) {
+                 std::complex<double> kvalue(realKImage(ikx, iky), imagKImage(ikx, iky));
+                 _ktab->kSet(kx, ky, kvalue);
+                 xxdbg<<"ikx,iky,kx,ky = "<<ikx<<','<<iky<<','<<kx<<','<<ky<<std::endl;
+                 xxdbg<<"kvalue = "<<kvalue<<std::endl;
+             }
+        }
+        _flux = kValue(Position<double>(0.,0.)).real();
+        dbg<<"flux = "<<_flux<<std::endl;
+
+        xcentroid = NAN;
+        ycentroid = NAN;
+    }
+
+    SBInterpolatedKImage::SBInterpolatedKImageImpl::~SBInterpolatedKImageImpl() {}
+
+    std::complex<double> SBInterpolatedKImage::SBInterpolatedKImageImpl::kValue(
+        const Position<double>& k) const
+    {
+        xdbg<<"evaluating kValue("<<k.x<<","<<k.y<<")"<<std::endl;
+        xdbg<<"_maxk = "<<_maxk<<std::endl;
+        if (std::abs(k.x) > _maxk || std::abs(k.y) > _maxk) return std::complex<double>(0.,0.);
+        return _ktab->interpolate(k.x, k.y, *_kInterp);
+    }
+
+    Position<double> SBInterpolatedKImage::SBInterpolatedKImageImpl::centroid() const {
+        double flux = getFlux();
+        if (flux == 0.) throw std::runtime_error("Flux == 0.  Centroid is undefined.");
+        if (std::isnan(xcentroid)) {
+            /*  int x f(x) dx = (x conv f)|x=0 = int FT(x conv f)(k) dk
+                              = int FT(x) FT(f) dk
+                FT(x) is divergent, but really we want the first integral above to be
+                int(x f(x) dx, -L/2..L/2) since f(x) is formally periodic once it's put on a
+                grid.  So in the last integral, we really want FT(x if |x|<L/2 else 0),
+                which works out to
+                2 i ( kx L cos(kx L/2) - 2 sin(kx L/2)) sin (ky L/2) / kx^2 / ky.
+                Noting that kx L/2 = ikx pi, the cosines are -1^ikx and the sines are 0.
+                Of course, lim kx->0 sin(kx)/kx is 1 though, so that term survives.  Algebra
+                eventually reduces the above expression to what's in the code below.
+             */
+            double xsum(0.0), ysum(0.0);
+            for (int iky = -Ninitial/2; iky < Ninitial/2; iky++) {
+                if (iky == 0) continue;
+                ysum += std::pow(-1.0, iky) / iky * _ktab->kval(0, iky).imag();
```

Comment:
A minor efficiency point here.  Rather than `pow(-1.0, iky)`, you can just keep a variable that keeps track of this and update it each time through the loop.
```
int iky = -Ninitial/2
double sign = (iky % 2 == 0) ? 1.0 : -1.0;
for (; iky < Ninitial/2; iky++) {
    sign = -sign;
    ...
```